New home page login

Bohjaangles opened the pull request:
Create login button styled to look like reddit in navbar
 - this opens a modal that displays a form for logging in and an option to declare that registering is needed first
 - if button for register, the modal changes to the register form

homepage was already pretty much made though is pretty blank less some minor layout

Bohjaangles committed:
add button for login, add login file for modal contents

Bohjaangles committed:
add login component that is a form that takes in email and password, and surrently outputs that info into the console. put that form into a modal that is triggered by a button labeled as -- login -- on the homepage navbar

Bohjaangles committed:
add a register option to my login component, add state to track whether user is registered or not, and a button for users to declare if they are already registered or not

andrew-x commented on src/Login.tsx:
```diff
@@ -0,0 +1,99 @@
+  PasswordInput,
+} from '@mantine/core'
+
+interface FormValues {
```
why interface vs type? what are the differences between the two?

Bohjaangles replied:
To my understanding, Interfaces are jsut better tools than types. As Interfaces can do everything a type can do, only types have more restriction, as in you can't extend a type buyt you could an interface, also, interfaces can cover complex data structures while types I think can only define primitives and arrays, while interfaces can also define objects. I suppose in a project where you would want to make it so that a certain definition cannot be extended or added to would make a type the better tool, but outside of that interfaces all day

andrew-x replied:
So definitely right high level idea but slight correction. Types could actually be extended but you're not 'extending' as much as you are composing. For example `type c = a & b`. This type logic where you can calculate intersections and unions is actually very powerful for complex data types. Most of the time though, I would just stick with one or the other and it comes down to personal preference. I prefer types but recognize that interfaces make things cleaner and offers better inspection for more niche linting too.

Bohjaangles replied:
Got it I think, I suppose if just declaring anything other than an object, might as well use a type, and for objects and declarations that are subject to possible change down the line, go with interface.

andrew-x replied:
Yup makes sense

andrew-x commented on src/Login.tsx:
```diff
@@ -0,0 +1,99 @@
+            onChange={handleChange}
+            required
+          />
+          <Button onClick={handleChangeIsRegistered}>Not yet registered?</Button>
```
these buttons aren't appearing correctly on the page, the conflict between mantine and tailwind hasn't been resolved yet. As a hint you need to update a setting in the tailwind config

Bohjaangles replied:
hmm, looking through the tailwind docs, the options looked like using either the important: true or prefix: 'tw-', and having tried both, I was able to get the tw- prefix to work, but it seems to only add extra work for me as I then had to refactor so that all my tw classnames had tw in front of them and the element just had no styling at all and was only visible on hover when I first made the change. I think I am missing something here but after spending about an hour and a half playing with it and looking at different things, I am ready to ask for additional help with this portion

andrew-x replied:
So the answer is actually tailwind has a preflight step in it's initialization that wipes out all existing HTML default styling. Mantine relies on some of this styling to function. To make the two work, you just need to disable tailwind preflight

andrew-x commented on src/Login.tsx:
```diff
@@ -0,0 +1,99 @@
+      {isRegistered === true ? 
+      <>
+        <Text>Log In</Text>
+        <form onSubmit={handleSubmit}>
```
try using something like react-hook-form or formik, it'll make validation a lot easier

Bohjaangles replied:
Well that was a bigger refactor than I expected. Ultimately decided to go with react-hook-form cause it played nicer with mantine lib.

andrew-x replied:
Good choice! Now there is a one key difference between how formik and react-hook-form checks state, did you run into that when you were looking into it?

Bohjaangles replied:
Honestly not really, both seemed to work pretty well out of the box, and so didn't look too deeply into how they managed state, as both spit out the correct string values in console when testing, the choice came down to which was easier to style with tbh. In looking into how they manage state, it looks like formik keeps track of state in it's formik component and so state is changed that way, and therefore everything inside the formik component re-renders whenever that state changes, where RHF tracks changes in the DOM el in a ref attribute so that state change only causes re-renders to happen in the local ele instead of the entire form. Is this correct?

andrew-x replied:
Sort of, basically Formik relies on the onChange parameter to listen for changes to update an internal state value. React-hook-forms is able to forego tracking internal state and basically only "grab" the value when the form is submitted. This only works for default HTML inputs though, if you have a custom input like a calendar picker or something along those lines, you would have to use onChange to track internal state still.

What that means in this case when you're using TextInput is instead of setting `value` and `onChange`, you can just call `{...register('field-name')}` to register the field with react-hook-forms and it'll take care of everything for you when onSubmit is called

andrew-x commented on src/Login.tsx:
```diff
@@ -0,0 +1,99 @@
+
+  return (
+    <div>
+      {isRegistered === true ? 
```
you can do `!!isRegistered` to cast it directly to a boolean value

Bohjaangles replied:
like that one!

andrew-x commented on src/App.tsx:
```diff
@@ -2,6 +2,7 @@ import { Button } from "@mantine/core"
 import Home from "./Home"
```
try setting the project up to accept absolute paths for imports, doesn't matter too much for these pages but will come into play more as the app gets bigger

Bohjaangles replied:
This one I thought would be a two second thing, but actually having issues with it. was basically trying vairations on this: https://dev.to/andrewezeani/how-to-create-absolute-imports-in-vite-react-app-a-step-by-step-guide-28co  seems like it should be working but I am getting module/type declaration errors. What would be the issue here? I mostly tried playing with vite.config, but maybe there is more to do that I don't know about?

andrew-x replied:
So there's two factors at play, one is with Vite knowing where to look but there's also a factor with typescript knowing where to look to. I think what will probably solve it is if you added baseUrl property into tsconfig

Bohjaangles replied:
Ah, gotcha, kk, yeah I got this one working. had to add 
"@/*": [
        "*"
      ]
to path. If I understand it, @/ anything is what is mapping to ./src, and the "*" means that anything in there can be mapped that way so long as src is parent. correct?

andrew-x replied:
I think so for this case (tbh don't quite remember correctly), might be different when using other frameworks like Next.js though. Gonna leave this as a "if it works it works" haha

Bohjaangles committed:
refactor login to use react-hook-form, minor ediminor edits to vite.config and tailwind config

andrew-x commented:
So you actually committed the Yarn cache files, can you remove them and update the .gitignore? Definitely don't want any temporary or cache files in a repo

Bohjaangles committed:
Delete .yarn dir

Accidently included that in last push. Updated gitignore to no longer include .yarn files

andrew-x commented on bin/App.js:
Bin files should also be removed since these are autogenerated

Bohjaangles replied:
done and done

andrew-x commented on tailwind.config.js:
```diff
@@ -1,5 +1,6 @@
 /** @type {import('tailwindcss').Config} */
 export default {
+  prefix: 'tw-',
```
better solution is to disable preflight for tailwind: https://tailwindcss.com/docs/preflight

Bohjaangles replied:
Yeah I disabled preflight in tailwind.config.js with a coreplugins:{ preflight: false,}, do you mean something else from what I did there?

andrew-x replied:
oh is it committed in the tailwind config file? I'm not seeing it I don't think

Bohjaangles replied:
huh, yeah, must be part of the merge conflict, had that change commited weeks ago lol

andrew-x commented on src/Login.tsx:
```diff
@@ -0,0 +1,144 @@
+      <>
+        <Text>Log In</Text>
+        <form onSubmit={handleSubmit(onSubmit)}>
+          <Controller
```
shouldn't need the controller, you should be able to subscribe the field to react-hook-forms directly

Bohjaangles replied:
haha, kk, this was a pretty big refactor, but I do understand how react hook forms works a bit better now I think. Think I got it now

andrew-x commented on src/Home.tsx:
```diff
@@ -11,31 +12,39 @@ import {
 
   return (
-    <div>
+    <div className='tw-'>
```
probably don't need this here

Bohjaangles replied:
Makes sense, everything is in appshell anyways, was a needless wrapper. kk

Bohjaangles committed:
updated jsconfig so that absolute paths can be used and updated a few files to follow that structure

Bohjaangles committed:
Merge branch 'newHomePage-Login' of github.com:Bohjaangles/redditClone into newHomePage-Login

Bohjaangles committed:
refactor login to remove controler wrapper, some other minor edits to needless divs

andrew-x commented:
Also got some merge conflicts if you can resolve it too

Bohjaangles committed:
compled to js

Bohjaangles committed:
added /bin to gitignore

Bohjaangles committed:
Delete bin dir from branch

andrew-x commented:
looks like `bin` files were committed again, might want to add to gitignore
